### DISCUSSION TITLE
why kernelSpec() return a copy of spec.Copy()

### DISCUSSION THREAD

**Discussion by albe19029** [cilium/ebpf#1394] [Ideas]
Good day. I have faced next problem.
When I am trying to load my CO-RE program with ebpf my program consume more then 140MB of memory even for small program.
More then 100 MB of this memory are allocated in kernelSpec().

It is OK, that 50 MB of memory are when we parsing vmlinux in loadKernelSpec(). But why do we make a copy of this data when return from kernelSpec()? I didn't manage to find a place where this data can be changed after the load
So the questing is it possible to use spec and not spec.Copy()

**Comment by florianl** [comment]
After your BTF operations, use [`btf.FlushKernelSpec()`](https://pkg.go.dev/github.com/cilium/ebpf@v0.13.2/btf#FlushKernelSpec) to remove any cached kernel type information and free memory.

**Comment by albe19029** [comment]
Yes, I do btf.FlushKernelSpec. But it don't resolve peak memory on process start. At start point I do have 2 copies of unparsed vmlinux information in memory, each of one occupies 50 MB.

**Comment by albe19029** [comment]
I have been working on version v0.10.0. Now I have checked master and found this commit : https://github.com/cilium/ebpf/commit/4ad6b8c1eed691c5628580c922c3534509d67a70

And with this commit even without btf.FlushKernelSpec() I see 72 Mb using of memory. Before 142 Mb.

So will just update to version v0.13.0 or higher and my life will be great.

Thanks for you great job with this project.
